docs/features/string-literals-data-section.md

```diff
@@ -0,0 +1,201 @@
+# String literals in data section
+
+This opt-in **experimental** Roslyn feature allows changing how string literals in C# programs are emitted into PE files (`.dll`/`.exe`).
+By default, string literals are emitted to the UserString heap which is limited to [2^24 bytes](https://github.com/dotnet/roslyn/issues/9852).
+When the limit is reached, the following compiler error is reported by Roslyn:
+
+```
+error CS8103: Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string literals.
+```
+
+By turning on [the feature flag](#configuration), string literals ([where possible](#eligible-string-literals)) are instead emitted as UTF-8 data into a different section of the PE file
+which does not have the same limit. The emit format is similar to [explicit u8 string literals][u8-literals].
+
+The feature is currently implemented only for C#, not VB.
+
+> [!WARNING]
+> This feature is currently experimental and can be changed or removed at any time.
+
+## Configuration
+
+The feature flag can take a non-negative integer threshold.
+Only string literals whose length (number of characters, not bytes) is over the threshold are emitted using the utf8 encoding strategy.
+If the flag is set, but no value is specified, the threshold defaults to 100.
+Specifying 0 means all string literals are considered for the feature.
+Specifying `off` as the value turns the feature off (this is the default).
+
+The feature flag can be specified on the command line like `/features:experimental-data-section-string-literals` or `/features:experimental-data-section-string-literals=20`,
+or in a project file in a `<PropertyGroup>` like `<Features>$(Features);experimental-data-section-string-literals</Features>` or `<Features>$(Features);experimental-data-section-string-literals=20</Features>`.
+
+## Eligible string literals
+
+A string literal is emitted with the utf8 encoding if and only if all of these are true:
+- it can be encoded as UTF-8, i.e., the following call does not throw:
+  ```cs
+  new System.Text.Encoding.UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true).GetBytes(literal)
+  ```
+- its `Length` is greater than the threshold specified by [the feature flag](#configuration).
+
+## Emit
+
+The standard string literal emit strategy emits `ldstr "string"`.
+That puts the string literal on the UserString heap of the PE file.
+
+The utf8 string literal encoding emit strategy emits `ldsfld` of a field in a generated class instead.
+
+For every string literal, a unique internal static class is generated which:
+- has name composed of `<S>` followed by a hex-encoded XXH128 hash of the string,
+- lives in the global namespace (for consistency with other compiler-generated types),
+- has one `.data` field which is emitted the same way as for [explicit u8 string literals][u8-literals],
+- has one `string` field which is initialized in a static constructor of the class.
```

Comment:
Are we going to apply `beforefieldinit` flag to the class?

Reply:
I don't think there is an observable difference between applying it or not (with current runtime implementations). Applying the flag can allow runtimes to call the static constructor at any time before the static string field is accessed, otherwise it would be called exactly before the string field is accessed for the first time. So I think applying it is better because it can allow runtimes to optimize if necessary.